### PR TITLE
[aggregation] add ratePerSecond

### LINKED ISSUE #452: Support Rate/RatePerSecond aggregations
Heroic should support Rate and RatePerSecond (irate) aggregations. Heroic already has Delta and DeltaPerSecond aggregations that can be used as examples when implementing...

 

Calculating the rate will allow us to calculate and comprehend SLO's in a much clearer way since we are getting the "total number of failed requests" / "all requests" vs calculating this from rates that are already aggregated on instances. 

Its easier to say 100/100000 requests are failing vs 0.01/1.9 (calculated with the avg of the rates). Also summing up the requests produces NO precision lost or issues with low throughput services or newly created meters.

 

Prometheus descriptions...

https://github.com/prometheus/prometheus/blob/2bd510a63e48ac6bf4971d62199bdb1045c93f1a/docs/querying/functions.md#rate

https://github.com/prometheus/prometheus/blob/2bd510a63e48ac6bf4971d62199bdb1045c93f1a/docs/querying/functions.md#irate

How prometheus handles counter resets, https://github.com/prometheus/prometheus/blob/c6399296dcc059bbc496f40e31aa2c7b7a341951/promql/functions.go#L178

https://github.com/prometheus/prometheus/blob/c6399296dcc059bbc496f40e31aa2c7b7a341951/promql/testdata/functions.test#L94

### PR DESCRIPTION
**Pull request by dmichel1**
The rate per second aggregation calculates the rate per second between samples in a given extent.

Rate per second works with sampling. If a queries resolution is 1 minute and points are published every 30s each bucket will contain 2 points. The rate will be calculated by getting the difference between the two points values and dividing them by the time between the points.

Note: at least 2 points are required in order to calculate the rate. Otherwise the previous value is unknown.

It's intended to be used with cumulative counters. It will handle the counters being reset because of restarts, etc.

I reimplemented an earlier pull request #408 that implemented `ratePerSecond` similarly to how `deltaPerSecond` works which does not support sampling the data. I think sampling is a requirement for most aggregations otherwise a large number of timeseries and points will likely cause the frontend to crash or produce graphs that are hard to read.

This closes #452.

### CODE REVIEW

File: aggregation/simple/src/main/java/com/spotify/heroic/aggregation/simple/RatePerSecondInstance.kt [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+
+/*
+    Calculate the rate per second of cumulative counters. When the counter is reset
+    because of an instance restart the interval will be skipped.
+ */
+data class RatePerSecondInstance(
+    override val size: Long,
+    override val extent: Long
+) : BucketAggregationInstance<RatePerSecondInstance.RateBucket>(size,
```
Comment (dmichel1) [inline aggregation/simple/src/main/java/com/spotify/heroic/aggregation/simple/RatePerSecondInstance.kt]: @hexedpackets this is the part I read about the buckets needing to be thread safe https://github.com/spotify/heroic/blob/092c340c30460129840fb8511c825fd773822ca4/heroic-component/src/main/java/com/spotify/heroic/aggregation/BucketAggregationInstance.kt#L43

File: aggregation/simple/src/main/java/com/spotify/heroic/aggregation/simple/RatePerSecondInstance.kt [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+    }
+
+    data class RateBucket(override val timestamp: Long) : AbstractBucket(), DoubleBucket {
+        private var firstPoint: Point? = null
+        private var lastPoint: Point? = null
+        private val rate = AtomicDouble()
+
+        override fun updatePoint(key: Map<String, String>, p: Point) {
+            if (firstPoint == null) {
```
Comment (dmichel1) [inline aggregation/simple/src/main/java/com/spotify/heroic/aggregation/simple/RatePerSecondInstance.kt]: and i'm pretttty sure this is not thread safe.

Comment (hexedpackets) [reply aggregation/simple/src/main/java/com/spotify/heroic/aggregation/simple/RatePerSecondInstance.kt]: Hmm yeah, this requires having chronologically serialized data.

Comment (dmichel1) [reply aggregation/simple/src/main/java/com/spotify/heroic/aggregation/simple/RatePerSecondInstance.kt]: i went through the calling code - when bigtable calls `metricsConsumer.accept(readResult);`  the entire metric collection is present and in order. 

As far as I can tell when doing a for-each aggregation, only a single thread is working on a given bucket at a time. I looked both at the code and also tracing what thread was executing the bucket (`grpc-default-executor-0` which is also weird? as that thread shouldn't be doing this work?).

Comment (dmichel1) [reply aggregation/simple/src/main/java/com/spotify/heroic/aggregation/simple/RatePerSecondInstance.kt]: @hexedpackets let's review this and confirm when you are back.

**Comment by dmichel1** [comment]
I'm merging this after I rebased with the latest changes.

**Pull request merged**